rafx-visibility/src/frustum_culling/packed_bounding_sphere_chunk.rs

```diff
@@ -0,0 +1,367 @@
+        for index in 0..last_index {
+            query.spheres = &chunk.spheres[index];
+            query.bitmask = query.spheres.is_contained_by(planes);
+            query.try_push_visibility_result(0);
+            query.try_push_visibility_result(1);
+            query.handle_index += PackedBoundingSphere::NUM_PACKED;
+        }
+    }
+
+    // NOTE(dvd): Check the last 2 spheres.
+    query.handle_index = last_index * PackedBoundingSphere::NUM_PACKED;
+    query.spheres = &chunk.spheres[last_index];
+    query.bitmask = query.spheres.is_contained_by(planes);
+    query.try_push_visibility_result(0);
+    if len & 1 == 0 {
+        // NOTE(dvd): If len is even, the last packed sphere has 2 spheres in it.
+        query.try_push_visibility_result(1);
+    }
+}
+
+struct Query<'a> {
+    pub camera_position: Vec3,
+    pub spheres: &'a PackedBoundingSphere,
+    pub handle_index: usize,
+    pub bitmask: i32,
+    pub metadata: &'a [ObjectMetadata; PackedBoundingSphereChunk::MAX_LEN],
+    pub results: &'a mut VisibleObjects,
+}
+
+impl Query<'_> {
+    #[inline(always)]
+    fn try_push_visibility_result(
+        &mut self,
+        packed_index: usize,
+    ) {
+        if (self.bitmask & (1 << packed_index)) > 0 {
+            let object = self.metadata[self.handle_index + packed_index];
+            self.results.push(VisibilityResult::new(
+                object.handle,
+                object.id,
+                self.camera_position,
+                self.spheres.get(packed_index),
+            ));
+        }
+    }
+}
+
+#[repr(C)]
+#[derive(Default, Copy, Clone, Debug, PartialEq)]
+struct PackedBoundingSphere {
+    pub x1: f32,
+    pub x2: f32,
+    pub y1: f32,
+    pub y2: f32,
+    pub z1: f32,
+    pub z2: f32,
+    pub radius1: f32,
+    pub radius2: f32,
+}
+
+impl PackedBoundingSphere {
+    pub const NUM_PACKED: usize = 2;
+
+    pub fn set(
+        &mut self,
+        index: usize,
+        sphere: BoundingSphere,
+    ) {
+        PackedBoundingSphere::assert_index_valid(index);
+        if index == 0 {
+            self.x1 = sphere.position.x;
+            self.y1 = sphere.position.y;
+            self.z1 = sphere.position.z;
+            self.radius1 = sphere.radius;
+        } else {
+            self.x2 = sphere.position.x;
+            self.y2 = sphere.position.y;
+            self.z2 = sphere.position.z;
+            self.radius2 = sphere.radius;
+        };
+    }
+
+    pub fn get(
+        &self,
+        index: usize,
+    ) -> BoundingSphere {
+        PackedBoundingSphere::assert_index_valid(index);
+        return if index == 0 {
+            BoundingSphere {
+                position: Vec3::new(self.x1, self.y1, self.z1),
+                radius: self.radius1,
+            }
+        } else {
+            BoundingSphere {
+                position: Vec3::new(self.x2, self.y2, self.z2),
+                radius: self.radius2,
+            }
+        };
+    }
+
+    #[inline(always)]
```

Comment:
did you measure this to be an improvement?

Reply:
(the `#[inline(always)]` I mean)

Reply:
Yes, at least on my machine. 🤷